sklearn/utils/testing.py

```diff
@@ -393,6 +403,177 @@ def assert_raise_message(exception, message, function, *args, **kwargs):
+        # of the results of multiple checks.
+        return not_same_model
+    return False  # if the method does not exist.
+
+
+def assert_same_model(X, estimator1, estimator2, msg=None):
+    """Helper function to check if models are same"""
+    _assert_same_model_method('predict', X, estimator1, estimator2, msg)
+    _assert_same_model_method('transform', X, estimator1, estimator2, msg)
+    _assert_same_model_method('decision_function',
+                              X, estimator1, estimator2, msg)
+    _assert_same_model_method('predict_proba',
+                              X, estimator1, estimator2, msg)
+
+    assert_fitted_attributes_equal(estimator1, estimator2)
+
+
+def assert_not_same_model(X, estimator1, estimator2, msg=None):
+    """Helper function to check if models are different"""
+    t1 = _not_same_model_method('predict', X, estimator1, estimator2)
+    t2 = _not_same_model_method('transform', X, estimator1, estimator2)
+    t3 = _not_same_model_method('decision_function', X, estimator1, estimator2)
+    t4 = _not_same_model_method('predict_proba', X, estimator1, estimator2)
+
+    if msg is None:
+        msg = ("Models are equal. Both models returned the same results for"
+               " all the methods.")
+
+    any_method_differed = any((t1, t2, t3, t4))
+
+    if not any_method_differed:
+        # If all methods return similar results for both ests or if those
+        # methods are not present
+        assert_fitted_attributes_not_equal(estimator1, estimator2)
+
+
+def _attributes_equal(estimator1, estimator2):
+    """Helper function to check if fitted model attributes are equal or not
+
+    Raises AssertionError if the attributes are not equal.
+    """
+
+    # A list of attributes which are known to be inconsistent.
+    skip_attributes = ('embedding_', 'weights_')
+    est1_dict = estimator1.__dict__
+    est2_dict = estimator2.__dict__
+    # TODO Add more comparables
+    # TODO Support sparse matrices?
+    comparables = [list, np.ndarray, np.float64, float, int, 'str']
+
+    # Collect all attributes from both the models
+    for attr, value1 in est1_dict.items():
+        if ((not attr.endswith('_')) or attr.endswith('estimator_') or
+                attr.endswith('estimators_') or
+                type(value1) not in comparables or
+                attr in skip_attributes):
+            continue
+
+        # If the attr is not present in the 2nd estimator
+        if attr not in est2_dict:
+            return False, 'attr_not_in_est2', attr
+
+        value2 = est2_dict[attr]
+        value1_type = type(value1)
+        value2_type = type(value2)
+
+        # If the attr types differ between the two estimators
+        if value1_type is not value2_type:
+            return (False, 'attr_not_same_type',
+                    (value1, value1_type), (value2, value2_type))
+
+        # TODO Support sparse matrices?
+        # If the attribute is of list / ndarray type
```

Comment:
> Is there a reason why we need almost here? :)
>
> I thought we were supposed to make sure that the models are exactly 
> the same... and the reason why hash is not used is that hash can't be 
> used directly on lists or dicts.. (we could convert them to tuple and 
> then hash them but again this has to be done recursively...)
>
> Please correct me if I am wrong!
>
I have no idea where your comment went, but we can never expect float 
numbers to be equal, right?
And joblib has hashes that work on everything iirc. @ogrisel?